### PR TITLE
fix(ci): harden BMI guard traversal against transient node_modules paths

### PR DESCRIPTION
**Pull request by Katsiarynakavaleuskaya**
## Summary
Hardens the BMI anti-duplication guard traversal so generated/transient dependency directories are pruned before descent instead of being discovered by `Path.rglob()`.

## Goal
Fix the main CI failure in `tests/test_no_bmi_math_outside_core.py::test_no_bmi_formula_outside_core` caused by transient `frontend/node_modules/fsevents` traversal errors while preserving the BMI source invariant.

## Business reason
Main CI should fail only on real BMI invariant violations, not on transient generated dependency paths from frontend optional dependencies.

## Scope
- Replace raw repo-wide `Path.rglob("*.py")` traversal in the BMI guard with a safe `os.walk(..., topdown=True)` walker.
- Prune generated/transient dependency and cache directories before descending.
- Keep source traversal/read errors fail-closed except the existing xdist helper path, now scoped to `app/test_guard_whr_skip_temp.py`.
- Add deterministic regression tests for node_modules pruning, traversal errors, read errors, path-scoped transient filtering, and source scanning.

## Out of scope
- No changes to `core/bmi/*`.
- No changes to `bmi_core.py`.
- No app/runtime changes.
- No frontend runtime or dependency changes.
- No design contracts, OpenAPI, backend routes, CI workflow YAML, tokens, or generated mirrors.

## Files changed
- `tests/test_no_bmi_math_outside_core.py`
- `docs/review/PR_1799_FIXED_MAPPING.md`

## Root cause
The failure was caused by repository-wide guard traversal entering generated frontend dependency directories. A transient/broken optional dependency path under frontend/node_modules/fsevents caused pathlib traversal to raise FileNotFoundError before the guard reached file-level OSError handling. The fix prunes generated dependency directories before traversal and keeps BMI source invariants unchanged.

## Source of truth
- `AGENTS.md`
- `RUNBOOK_AGENT.md`
- `tests/AGENTS.md`
- `tests/test_no_bmi_math_outside_core.py`
- `tests/test_repo_policy_guards.py`

## Agent Execution Log
- `agent-coordinator`: PASS for PR opening after scope and role synthesis.
- `architecture-specialist`: PASS; BMI regexes and whitelist semantics preserved.
- `security-auditor`: PASS after traversal/read errors outside excluded/generated paths were made fail-closed.
- `qa-engineer-agent`: PASS; deterministic regression coverage added and post-open mapping/test coverage fixes verified.
- `bug-hunter`: PASS after fixing false-green risks in source read errors, basename-only temp filtering, and parser-invalid fixed-mapping content.

## Skill Execution Log
- `pulseplate-workflow`: loaded for startup workflow and coordinator-first policy.
- `pulseplate-gates`: loaded and bounded local gates executed.
- `pulseplate-premortem-risk-review`: ran on actual diff; initial generic basename-pruning concern fixed by narrowing excluded dirs and proving `app/build/*.py` remains scanned.
- `pulseplate-pr-review`: post-open review completed through role-agent re-review; no blocking scoped findings remain.

## Experiment Runner Evidence
Artifact: `artifacts/orchestration/experiments/results/exp-75a941efa308.json`

Status: generated but rejected in isolated oracle environment because allowlisted `python` did not have repo dependencies installed:

```text
ModuleNotFoundError: No module named 'fastapi'
```

Local `.venv` bounded checks below are the validation evidence for this hotfix.

## Lane Start Provenance
Packet: `artifacts/orchestration/task_packets/b7da55d9481e.json`

Post-open packet: `artifacts/orchestration/task_packets/aeeb1f8e76bc.json`

## Premortem
Decision: proceed with changes.

Initial premortem concern: generic basename pruning (`build`, `dist`, `coverage`) could hide future source. Fixed by narrowing the excluded set to transient dependency/cache/tooling directories and adding a regression that `app/build/real_source.py` is still yielded by the walker.

## Security Review
PASS. The guard does not add suppressions, skips, or runtime/dependency scope. Traversal errors outside excluded/generated directories re-raise, and read-time errors for real source files re-raise.

## Bug Hunter Pass
PASS. Regression tests cover:
- `node_modules` pruning before descent.
- `frontend/node_modules/fsevents`-style traversal errors do not crash.
- source traversal errors fail closed.
- real source files remain scanned.
- source read errors fail closed.
- transient helper exceptions are scoped to `app/test_guard_whr_skip_temp.py`.
- exact transient helper read-time races are still tolerated.

## Tests / bounded checks
Startup:

```text
python3 scripts/orchestration/check_preflight.py
PASS: All required SoT files present
PASS: worktrees/ not tracked
PASS: agent consistency check
```

```text
python3 scripts/orchestration/check_agent_consistency.py
OK: agent docs and files are consistent.
```

Focused gates:

```text
.venv/bin/python -m pytest -q tests/test_no_bmi_math_outside_core.py
...........................                                              [100%]
```

```text
.venv/bin/python -m pytest -q tests/test_no_bmi_math_outside_core.py tests/test_repo_policy_guards.py
.........................................                                [100%]
```

```text
.venv/bin/python -m mypy --no-incremental --cache-dir=/dev/null tests/test_no_bmi_math_outside_core.py
Success: no issues found in 1 source file
```

```text
make validate-changed
✅ Diff-based validation completed
```

```text
.venv/bin/python -m pytest -q tests/test_no_bmi_math_outside_core.py::test_no_bmi_formula_outside_core
.                                                                        [100%]
```

```text
python3 scripts/ci/check_pr_body_phase2_gates.py --pr-number 1799 --body "..."
phase2-pr-body-gates: canonical mapping artifact and PR body mirror passed.
```

```text
python3 scripts/ci/check_docs_phase1_gates.py --files docs/review/PR_1799_FIXED_MAPPING.md
phase1-docs-gates: passed.
```

```text
pre-commit run --all-files
black (format)...........................................................Passed
ruff (lint, local).......................................................Passed
backend tests (pytest, changed files)....................................Passed
ios syntax check (swift).................................................Passed
```

Host-python note:

```text
python -m pytest -q tests/test_no_bmi_math_outside_core.py
ModuleNotFoundError: No module named 'fastapi'
```

This was rerun successfully with `.venv/bin/python` per repo policy.

## Deferred / Follow-ups
None.

## Discussion Thread Pass
- [x] Discussion-thread pass completed
- [x] Fixed in commit mapping completed

### Fixed in Commit Mapping
Canonical artifact: `docs/review/PR_1799_FIXED_MAPPING.md`

No actionable review comments at PR open.

## Merge Readiness
Not merge-ready yet. Current-head CI, external/bot review disposition, review-thread disposition guard, strict merge-readiness wrapper, and wait-window are still required before any readiness claim.

## Rollback
Rollback is revert-only: revert this PR to restore the previous BMI guard traversal behavior. No runtime rollback required.

## DoD
- [x] Generated dependency directories pruned before traversal.
- [x] BMI regex/invariant logic unchanged.
- [x] Real repo Python files still scanned.
- [x] Broken `node_modules/fsevents`-like traversal errors do not crash scanner.
- [x] Source traversal/read errors fail closed.
- [x] No runtime/design/dependency/CI workflow files touched.
- [x] Bounded local checks passed.

## Commit breakdown
- `ec532b4a9 fix(ci): harden BMI guard traversal`
- `e3d1cf20b docs(review): add PR 1799 fixed mapping`
- `9473538d7 test(ci): cover scoped BMI guard temp skips`
- `e7dab778f docs(review): record PR 1799 bot disposition`
- `2fd318855 test(ci): document BMI guard helpers`
- `dc2717212 docs(review): map PR 1799 docstring warning`
- `91d9818e0 docs(review): map PR 1799 CodeRabbit review`

## Pre-push checklist
- [x] `pre-commit run --all-files`
- [x] pre-push hooks passed during `git push`
- [x] `git ls-files worktrees artifacts/agent_runs artifacts/orchestration artifacts/security_lab .venv node_modules coverage.xml .coverage` returned empty

## Post-merge checklist
- Sync local `main` with `origin/main`.
- Confirm main current-head health.
- Remove merged local/remote branch after PR merge.
- Prune stale worktree metadata if needed.

## AGENTS.md updates
Not required. This hotfix changes one guard implementation and does not change agent workflow or repo policy.

## Summary by CodeRabbit

* **Tests**
  * Strengthened repo scanning and guard checks to prevent unintended math/threshold drift, with stricter read-error handling and path-scoped transient exceptions; added targeted tests covering pruning, walk/read errors, and threshold filtering edge cases.

* **Documentation**
  * Added a detailed review artifact documenting fixes, verification steps, audit evidence, and merge-readiness confirmation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Katsiarynakavaleuskaya/PulsePlate/pull/1799?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/test_no_bmi_math_outside_core.py [hunk 1]
```diff
@@ -273,9 +322,172 @@ def test_no_bmi_formula_outside_core() -> None:
...
+
+    source_dir = tmp_path / "app"
+    source_dir.mkdir()
+    source_file = source_dir / "real_source.py"
+    source_file.write_text("VALUE = 1\n", encoding="utf-8")
+
+    yielded = {path.relative_to(tmp_path).as_posix() for path in _iter_repo_python_files(tmp_path)}
+
+    assert "app/real_source.py" in yielded
+    assert "frontend/node_modules/fsevents/forbidden.py" not in yielded
```
[the remark this answers is not in the document]
Comment (Katsiarynakavaleuskaya) [reply tests/test_no_bmi_math_outside_core.py:355]: Disposition: NOT-A-BUG

Evidence: current-head `tests/test_no_bmi_math_outside_core.py` already has explicit `-> None` return annotations on the listed new/modified tests, including `test_iter_repo_python_files_skips_node_modules` at line 325, multiline walker tests at lines 357 and 383, and scan/path-scope tests at lines 419, 441, 465, and 486. Local check also passed: `.venv/bin/python -m mypy --no-incremental --cache-dir=/dev/null tests/test_no_bmi_math_outside_core.py` -> `Success: no issues found in 1 source file`.

Recorded in canonical mapping artifact: `docs/review/PR_1799_FIXED_MAPPING.md` in commit `e7dab778f`.

**Comment by Katsiarynakavaleuskaya** [comment]
CodeRabbit pre-merge docstring warning disposition: FIXED.

Added docstrings for the two new helper functions in `tests/test_no_bmi_math_outside_core.py` in commit `2fd318855`, then recorded the bot warning in the canonical mapping artifact in commit `dc2717212`.

Evidence:
- `tests/test_no_bmi_math_outside_core.py:216`
- `tests/test_no_bmi_math_outside_core.py:221`
- `pre-commit run --all-files` -> Passed
- `docs/review/PR_1799_FIXED_MAPPING.md` -> includes the mapped FIXED disposition

File: tests/test_no_bmi_math_outside_core.py [hunk 2]
```diff
@@ -197,8 +212,41 @@ def _should_skip_threshold_check(rel_path: str) -> bool:
...
 
 
+def _path_has_excluded_scan_part(path: Path) -> bool:
+    """Return whether a path is inside a generated/transient scan directory."""
+    return any(part in EXCLUDED_SCAN_DIR_NAMES for part in path.parts)
+
+
+def _is_allowed_transient_read_error(rel_path: str) -> bool:
+    """Return whether read-time disappearance is an expected xdist helper race."""
+    return rel_path == _GUARD_WHR_SKIP_TEMP_REL_PATH
```
[the remark this answers is not in the document]
Comment (Katsiarynakavaleuskaya) [reply tests/test_no_bmi_math_outside_core.py]: Disposition: FIXED

Implemented the separator-normalization fix in `105d60e74` and recorded it in `docs/review/PR_1799_FIXED_MAPPING.md` in `30db612d8`.

Evidence:
- `_normalize_rel_path(...)` normalizes backslashes to `/`.
- `_is_allowed_transient_read_error(...)` now compares normalized paths.
- `_scan(...)` uses `.as_posix()` for repo-relative paths.
- Added regression: `test_transient_read_error_path_match_is_separator_normalized`.
- Local checks passed: `.venv/bin/python -m pytest -q tests/test_no_bmi_math_outside_core.py` -> 28 passed; mypy and pre-commit passed.
